internal/controller/gatewayclass.go

```diff
@@ -108,24 +132,96 @@ func (r *gatewayClassReconciler) Reconcile(ctx context.Context, request reconcil
 				Name:      request.Name,
 			}})
 		return reconcile.Result{}, nil
-	} else if err != nil {
-		// Error reading the object, so requeue the request.
-		return reconcile.Result{}, fmt.Errorf("failed to get gatewayclass %q: %w", request.Name, err)
 	}
 
-	// The gatewayclass is safe to process, so check if it's valid.
-	errs := validation.ValidateGatewayClass(ctx, r.client, gc, r.controller)
-	if errs != nil {
-		r.log.WithField("name", gc.Name).Error("invalid gatewayclass: ", errors.ParseFieldErrors(errs))
+	for _, gc := range controlledClasses.notAdmittedClasses() {
+		if r.statusUpdater != nil {
+			r.statusUpdater.Send(k8s.StatusUpdate{
+				NamespacedName: types.NamespacedName{Name: gc.Name},
+				Resource:       gatewayClassGVR,
+				Mutator: k8s.StatusMutatorFunc(func(obj interface{}) interface{} {
+					gc, ok := obj.(*gatewayapi_v1alpha1.GatewayClass)
+					if !ok {
+						panic(fmt.Sprintf("unsupported object type %T", obj))
+					}
+
+					copy := gc.DeepCopy()
+					return status.SetGatewayClassAdmitted(context.Background(), r.client, copy, false)
+				}),
+			})
+		} else {
+			// this branch makes testing easier by not going through the StatusUpdater.
+			copy := status.SetGatewayClassAdmitted(context.Background(), r.client, gc.DeepCopy(), false)
+
+			if err := r.client.Status().Update(context.Background(), copy); err != nil {
+				return reconcile.Result{}, fmt.Errorf("error updating status of gateway class %s: %v", copy.Name, err)
+			}
+		}
 	}
 
-	// Pass the new changed object off to the eventHandler.
-	r.eventHandler.OnAdd(gc)
-
-	if err := status.SyncGatewayClass(ctx, r.client, gc, errs); err != nil {
-		return reconcile.Result{}, fmt.Errorf("failed to sync gatewayclass %q status: %w", gc.Name, err)
+	if r.statusUpdater != nil {
+		r.statusUpdater.Send(k8s.StatusUpdate{
+			NamespacedName: types.NamespacedName{Name: controlledClasses.admittedClass().Name},
+			Resource:       gatewayClassGVR,
+			Mutator: k8s.StatusMutatorFunc(func(obj interface{}) interface{} {
+				gc, ok := obj.(*gatewayapi_v1alpha1.GatewayClass)
+				if !ok {
+					panic(fmt.Sprintf("unsupported object type %T", obj))
+				}
+
+				copy := gc.DeepCopy()
+				return status.SetGatewayClassAdmitted(context.Background(), r.client, copy, true)
+			}),
+		})
+	} else {
+		// this branch makes testing easier by not going through the StatusUpdater.
+		copy := status.SetGatewayClassAdmitted(context.Background(), r.client, controlledClasses.admittedClass().DeepCopy(), true)
+		if err := r.client.Status().Update(context.Background(), copy); err != nil {
+			return reconcile.Result{}, fmt.Errorf("error updating status of gateway class %s: %v", copy.Name, err)
+		}
 	}
-	r.log.WithField("name", gc.Name).Info("synced gatewayclass status")
+
+	r.eventHandler.OnAdd(controlledClasses.admittedClass())
```

Comment:
eventually we can drop passing the GC to the DAG, we really just need to update the status, but leaving it for now until I get a PR up for #3864